### PR TITLE
Bug 1399149, new Notifications topic

### PR DESCRIPTION
**Pull request by ahardin-rh**
https://bugzilla.redhat.com/show_bug.cgi?id=1399149

Preview Build: http://file.rdu.redhat.com/~ahardin/112816/notifications/dev_guide/notifications.html

### CODE REVIEW

File: dev_guide/event_notifications.adoc
Comment (ahardin-rh) [inline dev_guide/event_notifications.adoc]: @tiwillia What more can be added here?

**Comment by ahardin-rh** [comment]
@tiwillia @jawnsy PTAL at this first pass. What more can be added?

Also, what are possible use cases for building from source, deploying on the cluster, etc.? The user can do these things, but why do they want to? Can you help me fill in the gaps? Thanks!

File: dev_guide/event_notifications.adoc
Comment (jawnsy) [inline dev_guide/event_notifications.adoc]: Should we use the term "Deployment Configuration" instead? The use of `deploymentConfig` looks specific to the way we specify deployment configuration data in JSON/YAML format, but that's an implementation detail. In other words, I recommend referring to the concept rather than our specific implementation, which is narrower.

Consider linking to the appropriate doc as well, such as: https://docs.openshift.org/latest/architecture/core_concepts/deployments.html#deployments-and-deployment-configurations (aside: on this page, we refer to a `DeploymentConfiguration` object, but it's defined as `kind: DeploymentConfig` - either we should use a space "Deployment Configuration" when referring to the concept, or "DeploymentConfig" when referring to the object type)

File: dev_guide/event_notifications.adoc
Comment (jawnsy) [inline dev_guide/event_notifications.adoc]: As above, I suggest using "persistent volume claim" (capitalized or not, depending on our doc practices; will leave that to you)

File: dev_guide/event_notifications.adoc
Comment (jawnsy) [inline dev_guide/event_notifications.adoc]: It may be useful to explain how we determine the project creator - we use the "requester" field of the project. (Do we show/refer to that as "project owner" in our UI?)

File: dev_guide/event_notifications.adoc
Comment (jawnsy) [inline dev_guide/event_notifications.adoc]: This seems like it gets too far into the weeds with respect to our implementation. It's useful for other developers in the Online team, as well as our ops team, but not to users that simply want notifications.

Consider a more generic statement here, explaining that we continually monitor events across the cluster and issue notifications when problems are detected. It's also worth mentioning that we throttle notifications (so you won't get flooded with emails)

File: dev_guide/event_notifications.adoc
Comment (jawnsy) [inline dev_guide/event_notifications.adoc]: We could add screenshots of the UI

File: dev_guide/event_notifications.adoc
Comment (jawnsy) [inline dev_guide/event_notifications.adoc]: As discussed briefly on IRC, I suggest commenting out this section, as it's only relevant to OpenShift Online developers & operators. This applies to the following sections:

* event-notifications-building-from-source (only Online devs)
* deploying-on-a-cluster (only Online ops)
* event-notifications-adding-new-dependencies (Online devs)

This is good content to have, because it'll be needed once we merge this into origin. However, as `openshift/online` is a private repo, nobody will be able to do this, and users will not have access to `openshift-infra` (only our ops team has cluster admin)

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: Agreed, this is really in-depth and not meant for user consumption.

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: This section is also developer-only info, the whole section can be taken out.

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: Another developer-only section we can take out.

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: Sorry, this is actually incorrect now, we should have updated the README to reflect the change when we made it. Here is a quick draft revision:

The notifications feature uses a ConfigMap named 'openshift-online-notifications' in each namespace to store preferences. The preferences allow users to specify which resources they would like to receive notifications for. An interface is provided in the console to easily modify these preferences. The following is an example ConfigMap for notification preferences:
```
{
    "kind": "ConfigMap",
    "apiVersion": "v1",
    "metadata": {
        "name": "openshift-online-notifications",
        "namespace": "example",
    },
    "data": {
        "builds-enabled": "true",
        "deployments-enabled": "true",
        "storage-enabled": "true"
    }
}
```

I'm not sure how necessary the example is, as it will have to be updated in the future when we add more options.

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: If we are moving to more readable resource names, this should probably be "build config"

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: I think a screenshot of the UI and/or a description of how to navigate to it would be great by the section describing notification preferences.

**Review by tiwillia — CHANGES_REQUESTED** [review]
As jawnsy has stated, several sections can be removed. Additionally, the section about notification preferences is incorrect (my fault, sorry!).

**Comment by ahardin-rh** [comment]
@jawnsy @tiwillia Thanks for your feedback and your help setting up the test system! This is updated. PTAL :bowing_woman: 

http://file.rdu.redhat.com/~ahardin/112816/notifications/dev_guide/event_notifications.html

File: dev_guide/event_notifications.adoc
Comment (ahardin-rh) [inline dev_guide/event_notifications.adoc]: @jawnsy @tiwillia Should this section be renamed to "Manually Configuring Notifications" since the topic leading into this is about configuring via the web console? Is that an accurate renaming?

File: dev_guide/event_notifications.adoc
Comment (ahardin-rh) [inline dev_guide/event_notifications.adoc]: Note to self: Add a link to the section above, as this blurb mentions the interface

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: That sounds appropriate to me. That or maybe just "Preference Storage", since I don't think we want to be encouraging folks to manually modify the ConfigMap.

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: This paragraph references the old preferences annotation(s).

It is true that if the **ConfigMap** is not present, it is assumed that notifications are disabled. The email address used is the email address of the project owner/

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: There is one more that was missed here:
```
*_Stuck deployment_:  A stuck deployment is a deployment where the active replicas is less than the replicas defined by the deploymentConfig for a certain amount of time. 
```

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: Instead of 'events' here, we should probably use the word 'resources'. An 'event' is a type of resource that we actually don't pay any attention to for notifications. We look at the state of active resources and determine if they are in a failed state without the use of events.

I think calling the notifications feature 'event notifications' should be fine though.

**Review by tiwillia — CHANGES_REQUESTED** [review]
Just a few more things

**Comment by ahardin-rh** [comment]
@jawnsy @tiwillia Updated. PTAL :eyes: 

@adellape please peer review :mag_right:

File: dev_guide/event_notifications.adoc
Comment (ahardin-rh) [inline dev_guide/event_notifications.adoc]: @tiwillia Is this "certain amount of time" specified somewhere? How long to be considered stuck?

Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: Hrmm well, it is configurable. I don't think we've decided on a specific amount of time for it yet. Maybe something like "for an unreasonable amount of time" will do for now?

Comment (jawnsy) [inline dev_guide/event_notifications.adoc]: I suggest "for a significant amount of time" or "for a long period of time" as _unreasonable_ seems, well, unreasonable ;)

**Comment by ahardin-rh** [comment]
@tiwillia @jawnsy Updated. Thanks!
 http://file.rdu.redhat.com/~ahardin/112816/notifications/dev_guide/event_notifications.html

FYI @abhgupta

**Comment by tiwillia** [comment]
LGTM :pizza: :tada:

File: getting_started/beyond_the_basics.adoc
Comment (adellape) [inline getting_started/beyond_the_basics.adoc]: @ahardin-rh One extra set of `../` here.

File: dev_guide/event_notifications.adoc
Comment (adellape) [inline dev_guide/event_notifications.adoc]: @ahardin-rh Could convert this to YAML?

**Review by adellape — COMMENTED** [review]
Couple notes inline, looks good tho! I like this approach.

File: dev_guide/event_notifications.adoc
Comment (tiwillia) [inline dev_guide/event_notifications.adoc]: Here ya go:
```
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: openshift-online-notifications
  namespace: example
data:
  builds-enabled: 'true'
  deployments-enabled: 'true'
  storage-enabled: 'true'
```

**Comment by ahardin-rh** [comment]
@tiwillia @adellape Updated. Thanks!

**Comment by ahardin-rh** [comment]
This is now QE verified. 

@abhgupta you previously mentioned that you may want to hold before merging this. Are we okay to move forward?

**Comment by ahardin-rh** [comment]
@dinhxuanvu is still doing testing on this feature, so I am holding off on merging for now. 

This is an OpenShift Online-only feature that will be implemented around the 3.4 time-frame.

**Comment by ahardin-rh** [comment]
Quick Update: I should be able to get access to the environment next week so that i can update the screenshots to reflect the latest UI. This is targeted for the paid tier release (release date still TBD).

**Comment by ahardin-rh** [comment]
@tiwillia Thanks for getting me access to update the screenshots. This is updated now. When you get a moment, can you please give this a review and let me know if anything else has changed? :beers:

**Comment by tiwillia** [comment]
@ahardin-rh Just one nit-pick, we are still calling them "Event notifications". I fear that because openshift has a completely separate concept of [events](https://docs.openshift.com/enterprise/3.2/dev_guide/events.html), these notifications might be confused as email notifications for openshift events. In reality, these are notifications on failures that occur in the background, where a user wouldn't possibly know about them unless they were actively watching the console.

Can we change the wording to "Failure notifications" rather than "Event notifications"?

**Comment by ahardin-rh** [comment]
@tiwillia Yes, agreed. This is now updated across pages.

**Comment by jawnsy** [comment]
"Failure Notifications" sounds too restrictive for future features we might want, though. I can see it being useful to warn you of stuff like "you're near your capacity" where it might not be causing you problems _now_ but is likely to cause you grief later. Is the name something we can change later?

**Comment by tiwillia** [comment]
LGTM now @ahardin-rh  🎉 Thanks!

**Comment by ahardin-rh** [comment]
@jawnsy Good point. The UI refers to it simply as "Notifications", so that is also an option. However, it is not very descriptive. We can always change the name later, but that may confuse users.

**Comment by tiwillia** [comment]
@ahardin-rh @jawnsy I'm okay with moving to "notifications" rather than "failure notifications". We do describe the feature in the docs as notifying on "various failures" so it should be fairly evident what the notifications are.

**Comment by ahardin-rh** [comment]
@tiwillia @jawnsy All updated: http://file.rdu.redhat.com/~ahardin/112816/notifications/dev_guide/notifications.html

**Comment by tiwillia** [comment]
Looks great to me @ahardin-rh

**Review by tiwillia — APPROVED** [review]

**Comment by ahardin-rh** [comment]
@jawnsy @tiwillia In the UI, we will need to update the "Learn more" link to point to the updated (renamed) path: https://docs.openshift.com/online/dev_guide/notifications.html

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../dev_guide/notifications.adoc#dev-guide-notifications[Notifications]
|New topic on the Notifications feature, which allows you to opt in to receive email notifications
about various failures, including dead or failed deployments, dead builds, and dead or failed persistent volume claims (PVCs) for each of your projects.
%

**Pull request merged**
